reco_utils/recommender/fbt/fbt.py

```diff
@@ -0,0 +1,311 @@
+        self._is_fit = False
+
+    def __repr__(self):
+        """Make a friendly, human-readable object summary string."""
+        return (
+            f"{self.__class__.__name__}(user_colname={self.col_user}, "
+            f"item_colname={self.col_item}, "
+            f"score_colname={self.col_score}, "
+            f"num_recos={self.num_recos}, "
+        )
+
+    def _check_dataframe(self, df,
+                         expected_columns=None):
+        """Verify input is a data frame and has expected columns.
+
+        Checks if the input is a DataFrame object and if so,
+        if the mandatory columns are present. Also checks for
+        duplicate rows.
+        """
+        if not isinstance(df, pd.DataFrame):
+            raise TypeError((
+                "Input argument must be a pandas DataFrame. "
+                f"Instead got {type(df)}!"))
+
+        if not expected_columns:
+            expected_columns = [self.col_user, self.col_item]
+        for col in expected_columns:
+            if col not in df.columns:
+                raise KeyError((
+                    f"Column {col} not found in DataFrame!"))
+
+        # check there are no duplicate rows in dataframe
+        if df.groupby(expected_columns).size().max() > 1:
+            raise ValueError("Duplicate rows found!!")
+
+    def fit(self, df):
+        """Fit the FBT recommender using an input DataFrame.
+
+        Fitting of model involves computing a item-item co-occurrence
+        matrix: how many people bought watched a pair of movies?
+
+        Args:
+            df (pd.DataFrame): DataFrame of users and items
+
+        Returns:
+        self
+            Fitted estimator.
+        """
+        logger.info("Check dataframe is of the type, schema we expect")
+        expected_cols_df = [self.col_user,
+                            self.col_item,
+                            f'{self.col_item}_name']
+        self._check_dataframe(df, expected_columns=expected_cols_df)
+
+        # copy the DataFrame to avoid modification of the input
+        select_columns = [self.col_user, self.col_item]
+        temp_df = df[select_columns].copy()
+
+        logger.info("De-duplicating the user-item counts")
+        temp_df = temp_df.drop_duplicates(
+                [self.col_user, self.col_item], keep="last"
+        )
+
+        # To compute co-occurrence, key piece is the self-join
+        cooccurrence_df = (
+            temp_df
+            .merge(temp_df, on=self.col_user,
+                   how='inner', suffixes=['', '_paired'])
+        )
+        # similarity score is defined by how many distinct
+        # users interacted with the same pair of items
+        sim_df = (
+            cooccurrence_df
+            .groupby([self.col_item,
+                     f'{self.col_item}_paired'])[self.col_user]
+            .nunique()
+            .reset_index(drop=False)
+            .rename(columns={self.col_user: self.col_score})
+        )
```

Comment:
in this [function](https://github.com/microsoft/recommenders/blob/main/reco_utils/recommender/sar/sar_singlenode.py#L159) we compute the coocurrence using numpy and scipy, changing to this format should give you a boost in performace

Reply:
Will update.

Reply:
So after updating the functions to use lumpy and scipy, the code for FBT looks very similar to SAR in many places. To minimize code replication, I was thinking of updating the SAR code directly to have FBT as part of it. changes going forward will be on `sar_singlenode.py` and there will be an example notebook showing that use case. Does that work @miguelgfierro @anargyri ? Who knows sar code the best for me to ping for advice? I can outline what I'm changing in the SAR code in a different comment.

Reply:
Tagging @gramhagen